core/network/src/discovery.rs

```diff
@@ -224,50 +237,58 @@ where
-							if closer_peers.is_empty() && self.num_connections != 0 {
-								warn!(target: "sub-libp2p", "Libp2p => Random Kademlia query has yielded empty \
-									results");
-							}
-						}
-						KademliaOut::GetValueResult(res) => {
-							let ev = match res {
-								GetValueResult::Found { results } => {
-									let results = results
-											.into_iter()
-											.map(|r| (r.key, r.value))
-											.collect();
-
-									DiscoveryOut::ValueFound(results)
-								}
-								GetValueResult::NotFound { key, .. } => {
-									DiscoveryOut::ValueNotFound(key)
-								}
-							};
-							return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
-						}
-						KademliaOut::PutValueResult(res) => {
-							let ev = match res {
-								PutValueResult::Ok{ key, .. } =>  {
-									DiscoveryOut::ValuePut(key)
-								}
-								PutValueResult::Err { key, .. } => {
-									DiscoveryOut::ValuePutFailed(key)
+				Async::Ready(NetworkBehaviourAction::GenerateEvent(ev)) => match ev {
+					KademliaEvent::UnroutablePeer { peer, .. } => {
+						let ev = DiscoveryOut::UnroutablePeer(peer);
+						return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
+					}
+					KademliaEvent::RoutingUpdated { peer, .. } => {
+						let ev = DiscoveryOut::Discovered(peer);
+						return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
+					}
+					KademliaEvent::GetClosestPeersResult(res) => {
+						match res {
+							Err(e) => {
+								trace!(target: "sub-libp2p", "Kademlia query for closest peers failed: {:?}", e);
+							},
+							Ok(ok) => {
+								trace!(target: "sub-libp2p", "Libp2p => Query for {:?} yielded {:?} results",
+									ok.key, ok.peers.len());
+								if ok.peers.is_empty() && self.num_connections != 0 {
+									warn!(target: "sub-libp2p", "Libp2p => Random Kademlia query has yielded empty \
+										results");
 								}
-							};
-							return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
+							}
 						}
-						// We never start any other type of query.
-						KademliaOut::GetProvidersResult { .. } => {}
+					}
+					KademliaEvent::GetRecordResult(res) => {
+						let ev = match res {
+							Ok(ok) => {
+								let results = ok.records
+									.into_iter()
+									.map(|r| (r.key, r.value))
+									.collect();
+
+								DiscoveryOut::ValueFound(results)
+							}
+							Err(e) => {
+								DiscoveryOut::ValueNotFound(e.into_key())
+							}
+						};
+						return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
+					}
+					KademliaEvent::PutRecordResult(res) => {
+						let ev = match res {
+							Ok(ok) => DiscoveryOut::ValuePut(ok.key),
+							Err(e) => {
+								DiscoveryOut::ValuePutFailed(e.into_key())
+							}
+						};
+						return Async::Ready(NetworkBehaviourAction::GenerateEvent(ev));
+					}
+					// We never start any other type of query.
+					e => {
+						trace!(target: "sub-libp2p", "Ignored Kademlia event: {:?}", e)
```

Comment:
Let's make this more verbose:

```suggestion
						warn!(target: "sub-libp2p", "Libp2p => Unexpected Kademlia event: {:?}", e)
```

Reply:
That implies some other changes though, since not all events that are currently ignored are unexpected. See the changes I made in: https://github.com/paritytech/substrate/pull/3076/commits/f5259e24ee42c6c4b72510fe1ee51b11a713cae1.